README.txt
  This program is a calculator that does
approximate integration. The function integrated,
the end points, which of five methods of
approximate integration to use (Left-hand rule,
Right-hand rule, Midpoint rule, Trapezoid rule,
and Simpson's rule), and how many divisions to
split the integrated region into.

  This program is released into the Public Domain
through the Creative Commons Zero 1.0 Universal
Public Domain Dedication (see LICENSE.txt file)

VERSION SPECIFIC INFORMATION

VERSIONS 1.x:
  Any version beginning with the number 1 of this
program requires a calc.cpp file that it will
write to. If one does not exist, it should create
one itself. This external file will not be needed
when version 2 releases.

VERSIONS 1.0-1.1
  Versions 1.0 and 1.1 of this program only run
on GNU/Linux systems that have the GNU C++
compiler run with "g++". They can be adapted for
other systems but the system() commands at the
end must be changed to whatever commands are used
to run the compiler and the compiled calc.cpp
program. This problem was fixed in version 1.2.


changelog.txt
Version 1.0 (unnumbered)
January 19 2023

Version 1.1
January 20 2023
Added support for mathematical constants e, pi, and tau

Version 1.2
January 21 2023
Removed declarations of e and pi since they're already in cmath
Made compiling and running new program manual to work across systems

February 2 2023
Fixed missing semicolon


calculator2.cpp
// THIS PROGRAM RELEASED INTO THE PUBLIC DOMAIN UNDER CREATIVE COMMONS ZERO 1.0 UNIVERSAL PUBLIC DOMAIN DEDICATION
// VERSION: 2.0-dev1
#include <iostream>
#include <cmath>
using namespace std;
//M_E AND M_PI ARE IN CMATH
//const double M_E=2.718281828459045235360;
//const double M_PI=3.14159265358979323846264338327950;
const double M_TAU=6.28318530717958647692528;
//LEFT HAND RULE
template <typename Func>double LHR(double a,double b,int n, Func func){
  double dx=(b-a)/n;
  double output;
  for(int i=0; i<n; i++){
    output += func(a+(i*dx));
  }
  output *= dx;
  return output;
}
//RIGHT HAND RULE
template <typename Func>double RHR(double a,double b,int n, Func func){
  double dx=(b-a)/n;
  double output;
  for(int i=1; i<=n; i++){
    output += func(a+(i*dx));
  }
  output *= dx;
  return output;
}
//MIDPOINT RULE
template <typename Func>double MR(double a,double b,int n, Func func){
  double dx=(b-a)/n;
  double output;
  for(int i=0; i<n; i++){
    output += func(a+((i+0.5)*dx));
  }
  output *= dx;
  return output;
}
//TREAPEZOID RULE
template <typename Func>double TR(double a,double b,int n, Func func){
  double dx=(b-a)/n;
  double output;
  for(int i=0; i<n; i++){
    output += (func(a+(i*dx))+func(a+((i+1)*dx)));
  }
  output *= (dx/2);
  return output;
}
//SIMPSON RULE
template <typename Func>double SR(double a,double b,int n, Func func){
  return (TR(a,b,n,func)/3)+(MR(a,b,n,func)*(2/3));
}

//EVALUATOR
struct Expression{
  char op;
  Expression* l;
  Expression* r;
}

Expression parse(string func){
  // OH NO THE PARSER
}

double evalaute (Expression func){
  //EVALUTE PARSED EXPRESSION
}

int main(){
  double a,b; //Endpoints of integration
  int n; //Number of sections
  string func=""; //Where the function will be
  char rule='\0'; //Which rule is used
  double integral; //Final integral
  cout << "Enter the function you want an approximate integral of, in C++ code. Use the variable name x (lowercase).\n(using the cmath library is allowed)\n";
  cout << "Three mathematical constants can be used: M_E (e, 2.71828...), M_PI (pi, 3.14159...), and M_TAU (tau = 2*pi, or 6.28318...)\n";
  cout << "The function will be placed into another file looking like this:\n\n";
  cout << "double func(double x){\n  double out=%s;\n  return out;\n}\n\n";
  cout << "Where %s contains the text of your function.\n";
  getline(cin, func);
  cout << "Enter the start and end points for your approximate integration.\n";
  cin >> a >> b;
  double temp; //To make sure a<b
  if(a>b){
    temp=a;
    a=b;
    b=temp;
  }
  cout << "Which integration method to use?\n";
  cout << "(L)eft-hand rule, (R)ight-hand rule, (M)idpoint rule, (T)rapezoid rule, or (S)impson's rule?\n";
  cin >> rule;
  cout << "How many intervals to split this into? (Must be even if Simpson's rule chosen)\n";
  cin >> n;
  switch(rule){
    case 'L': //LEFT HAND RULE
    case 'l':
      integral=LHR(a, b, n, func);
      break;
    case 'R':
    case 'r':
      integral=RHR(a,b,n,func);
      break;
    case 'M':
    case 'm':
      integral=MR(a,b,n,func);
      break;
    case 'T':
    case 't':
      integral=TR(a,b,n,func);
      break;
    case 'S':
    case 's':
      if(fmod(n,2)){
        cerr << "The number of divisions must be even for Simpson's rule.\n";
        return 1;
      }
      integral=SR(a,b,n/2,func);
      break;
    default:
      cerr << "Not a valid input.\n";
      return 1;
  }
  cout << integral << endl;
  return 0;
}


calculator.cpp
// THIS PROGRAM RELEASED INTO THE PUBLIC DOMAIN UNDER CREATIVE COMMONS ZERO 1.0 UNIVERSAL PUBLIC DOMAIN DEDICATION
// VERSION: 1.2
#include <iostream>
#include <fstream>
#include <cstdlib>
#include <cmath>
using namespace std;
ofstream out;
int main(){
    double a,b; //Endpoints of integration
    int n; //Number of sections
    string func=""; //Where the function will be
    char rule='\0'; //Which rule is used
    cout << "Enter the function you want an approximate integral of, in C++ code. Use the variable name x (lowercase).\n(using the cmath library is allowed)\n";
    cout << "In addition to the mathematical constants in the cmath library, you can also use M_TAU (tau = 2*pi, or 6.28318...)\n";
    cout << "The function will be placed into another file looking like this:\n\n";
    cout << "double func(double x){\n  double out=%s;\n  return out;\n}\n\n";
    cout << "Where %s contains the text of your function.\n";
    getline(cin, func);
    cout << "Enter the start and end points for your approximate integration.\n";
    cin >> a >> b;
    double temp; //To make sure a<b
    if(a>b){
        temp=a;
        a=b;
        b=temp;
    }
    cout << "Which integration method to use?\n";
    cout << "(L)eft-hand rule, (R)ight-hand rule, (M)idpoint rule, (T)rapezoid rule, or (S)impson's rule?\n";
    cin >> rule;
    cout << "How many intervals to split this into? (Must be even if Simpson's rule chosen)\n";
    cin >> n;
    out.open("calc.cpp");
    switch(rule){
        case 'L': //Left-hand rule
        case 'l':
        out << "#include <iostream>\n";
        out << "#include <cmath>\n";
        out << "using namespace std;\n";
        out << "const double M_TAU=6.28318530717958647692528;\n";
        out << "double func(double x){\n";
        out << "  double out=" << func << ";\n";
        out << "  return out;\n";
        out << "}\n";
        out << "int main(){\n";
        out << "  double a=" << a << ";\n";
        out << "  double b=" << b << ";\n";
        out << "  int n=" << n << ";\n";
        out << "  double dx=(b-a)/n;\n";
        out << "  double output;\n";
        out << "  for(int i=0; i<n; i++){\n";
        out << "    output += func(a+(i*dx));\n";
        out << "  }\n";
        out << "  output *= dx;\n";
        out << "  cout << output << endl;\n";
        out << "  return 0;\n";
        out << "}\n";
        break;
        case 'R': //Right-hand rule
        case 'r':
        out << "#include <iostream>\n";
        out << "#include <cmath>\n";
        out << "using namespace std;\n";
        out << "const double M_TAU=6.28318530717958647692528;\n";
        out << "double func(double x){\n";
        out << "  double out=" << func << ";\n";
        out << "  return out;\n";
        out << "}\n";
        out << "int main(){\n";
        out << "  double a=" << a << ";\n";
        out << "  double b=" << b << ";\n";
        out << "  int n=" << n << ";\n";
        out << "  double dx=(b-a)/n;\n";
        out << "  double output;\n";
        out << "  for(int i=1; i<=n; i++){\n";
        out << "    output += func(a+(i*dx));\n";
        out << "  }\n";
        out << "  output *= dx;\n";
        out << "  cout << output << endl;\n";
        out << "  return 0;\n";
        out << "}\n";
        break;
        case 'M': //Midpoint rule
        case 'm':
        out << "#include <iostream>\n";
        out << "#include <cmath>\n";
        out << "using namespace std;\n";
        out << "const double M_TAU=6.28318530717958647692528;\n";
        out << "double func(double x){\n";
        out << "  double out=" << func << ";\n";
        out << "  return out;\n";
        out << "}\n";
        out << "int main(){\n";
        out << "  double a=" << a << ";\n";
        out << "  double b=" << b << ";\n";
        out << "  int n=" << n << ";\n";
        out << "  double dx=(b-a)/n;\n";
        out << "  double output;\n";
        out << "  for(int i=0; i<n; i++){\n";
        out << "    output += func(a+((i+0.5)*dx));\n";
        out << "  }\n";
        out << "  output *= dx;\n";
        out << "  cout << output << endl;\n";
        out << "  return 0;\n";
        out << "}\n";
        break;
        case 'T': //Trapezoid rule
        case 't':
        out << "#include <iostream>\n";
        out << "#include <cmath>\n";
        out << "using namespace std;\n";
        out << "const double M_TAU=6.28318530717958647692528;\n";
        out << "double func(double x){\n";
        out << "  double out=" << func << ";\n";
        out << "  return out;\n";
        out << "}\n";
        out << "int main(){\n";
        out << "  double a=" << a << ";\n";
        out << "  double b=" << b << ";\n";
        out << "  int n=" << n << ";\n";
        out << "  double dx=(b-a)/n;\n";
        out << "  double output;\n";
        out << "  for(int i=0; i<n; i++){\n";
        out << "    output += (func(a+(i*dx))+func(a+((i+1)*dx)));\n";
        out << "  }\n";
        out << "  output *= (dx/2);\n";
        out << "  cout << output << endl;\n";
        out << "  return 0;\n";
        out << "}\n";
        break;
        case 'S': //Simpson's rule
        case 's':
        if(fmod(n,2)){
            cout << "The number of divisions must be even for Simpson's rule.\n";
            return 1;
        }
        out << "#include <iostream>\n";
        out << "#include <cmath>\n";
        out << "using namespace std;\n";
        out << "const double M_TAU=6.28318530717958647692528;\n";
        out << "double func(double x){\n";
        out << "  double out=" << func << ";\n";
        out << "  return out;\n";
        out << "}\n";
        out << "int main(){\n";
        out << "  double a=" << a << ";\n";
        out << "  double b=" << b << ";\n";
        out << "  int n=" << n/2 << ";\n"; //A weighted average of Midpoint and Trapezoid with n/2 divisions is equal to Simpson with n divisions
        out << "  double dx=(b-a)/n;\n";
        out << "  double output;\n";
        out << "  for(int i=0; i<n; i++){\n";
        out << "    output += dx*(func(a+(i*dx))+func(a+((i+1)*dx)))/2;\n";
        out << "    output +=dx*func(a+((i+0.5)*dx))*2;\n";
        out << "  }\n";
        out << "  output /= 3;\n";
        out << "  cout << output << endl;\n";
        out << "  return 0;\n";
        out << "}\n";
        break;
        default:
        cout << "Not a valid input.\n";
        return 1;
    }
    out.close();
    cout << "Please compile and run 'calc.cpp' to obtain the answer.\n";
    return 0;
}
